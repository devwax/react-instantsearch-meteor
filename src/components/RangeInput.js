import PropTypes from 'prop-types';
import React, { Component } from 'react';
import translatable from '../core/translatable';
import classNames from './classNames.js';

const cx = classNames('RangeInput');

export class RawRangeInput extends Component {
  static propTypes = {
    canRefine: PropTypes.bool.isRequired,
    precision: PropTypes.number.isRequired,
    translate: PropTypes.func.isRequired,
    refine: PropTypes.func.isRequired,
    min: PropTypes.number,
    max: PropTypes.number,
    currentRefinement: PropTypes.shape({
      min: PropTypes.number,
      max: PropTypes.number,
    }),
  };

  static defaultProps = {
    currentRefinement: {},
  };

  static contextTypes = {
    canRefine: PropTypes.func,
  };

  constructor(props) {
    super(props);

    this.state = this.normalizeStateForRendering(props);
  }

  componentWillMount() {
    if (this.context.canRefine) {
      this.context.canRefine(this.props.canRefine);
    }
  }

  componentWillReceiveProps(nextProps) {
    // In react@16.0.0 the call to setState on the inputs trigger this lifecycle hook
    // because the context has changed (for react). I don't think that the bug is related
    // to react because I failed to reproduce it with a simple hierarchy of components.
    // The workaround here is to check the differences between previous & next props in order
    // to avoid to override current state when values are not yet refined. In the react documentation,
    // they DON'T categorically say that setState never run componentWillReceiveProps.
    // see: https://reactjs.org/docs/react-component.html#componentwillreceiveprops

    if (
      nextProps.canRefine &&
      (this.props.canRefine !== nextProps.canRefine ||
        this.props.currentRefinement.min !== nextProps.currentRefinement.min ||
        this.props.currentRefinement.max !== nextProps.currentRefinement.max)
    ) {
      this.setState(this.normalizeStateForRendering(nextProps));
    }

    if (
      this.context.canRefine &&
      this.props.canRefine !== nextProps.canRefine
    ) {
      this.context.canRefine(nextProps.canRefine);
    }
  }

  onSubmit = e => {
    e.preventDefault();

    this.props.refine({
      min: this.state.from,
      max: this.state.to,
    });
  };

  normalizeStateForRendering(props) {
    const { canRefine, min: rangeMin, max: rangeMax } = props;
    const { min: valueMin, max: valueMax } = props.currentRefinement;

    return {
      from:
        canRefine && valueMin !== undefined && valueMin !== rangeMin
          ? valueMin
          : '',
      to:
        canRefine && valueMax !== undefined && valueMax !== rangeMax
          ? valueMax
          : '',
    };
  }

  normalizeRangeForRendering({ canRefine, min, max }) {
    const hasMin = min !== undefined;
    const hasMax = max !== undefined;

    return {
      min: canRefine && hasMin && hasMax ? min : '',
      max: canRefine && hasMin && hasMax ? max : '',
    };
  }

  render() {
    const { from, to } = this.state;
    const { precision, translate, canRefine } = this.props;
    const { min, max } = this.normalizeRangeForRendering(this.props);
    const step = 1 / Math.pow(10, precision);

    return (
      <form
        {...cx('root', !canRefine && 'noRefinement')}
        onSubmit={this.onSubmit}
      >
        <fieldset disabled={!canRefine} {...cx('fieldset')}>
          <label {...cx('labelMin')}>
            <input
              {...cx('inputMin')}
              type="number"
              min={min}
              max={max}
              value={from}
              step={step}
              placeholder={min}
              onChange={e => this.setState({ from: e.currentTarget.value })}
            />
          </label>
          <span {...cx('separator')}>{translate('separator')}</span>
          <label {...cx('labelMax')}>
            <input
              {...cx('inputMax')}
              type="number"
              min={min}
              max={max}
              value={to}
              step={step}
              placeholder={max}
              onChange={e => this.setState({ to: e.currentTarget.value })}
            />
          </label>
          <button {...cx('submit')} type="submit">
            {translate('submit')}
          </button>
        </fieldset>
      </form>
    );
  }
}

export default translatable({
  submit: 'ok',
  separator: 'to',
})(RawRangeInput);
