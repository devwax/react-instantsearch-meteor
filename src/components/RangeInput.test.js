/* eslint-env jest, jasmine */

import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import RangeInput, { RawRangeInput } from './RangeInput';

Enzyme.configure({ adapter: new Adapter() });

describe('RangeInput', () => {
  const shallowRender = (props = {}, context = {}) => {
    const defaultProps = {
      currentRefinement: {
        min: undefined,
        max: undefined,
      },
      canRefine: true,
      precision: 0,
      refine: () => {},
      min: undefined,
      max: undefined,
    };

    return shallow(<RangeInput {...defaultProps} {...props} />, { context });
  };

  it('render with translations', () => {
    const props = {
      translations: {
        submit: 'SUBMIT',
        separator: 'SEPARATOR',
      },
    };

    const component = shallowRender(props).shallow();

    expect(component).toMatchSnapshot();
  });
});

describe('RawRangeInput', () => {
  const shallowRender = (props = {}, context = {}) => {
    const defaultProps = {
      currentRefinement: {
        min: undefined,
        max: undefined,
      },
      canRefine: true,
      precision: 0,
      refine: () => {},
      translate: x => x,
      min: undefined,
      max: undefined,
    };

    return shallow(<RawRangeInput {...defaultProps} {...props} />, { context });
  };

  it('render with empty values', () => {
    const props = {};

    const component = shallowRender(props);

    expect(component).toMatchSnapshot();
  });

  it('render with empty values when refinement is equal to min / max', () => {
    const props = {
      currentRefinement: {
        min: 0,
        max: 500,
      },
      min: 0,
      max: 500,
    };

    const component = shallowRender(props);

    expect(component).toMatchSnapshot();
  });

  it('render with refinement', () => {
    const props = {
      currentRefinement: {
        min: 10,
        max: 490,
      },
    };

    const component = shallowRender(props);

    expect(component).toMatchSnapshot();
  });

  it('render with min / max', () => {
    const props = {
      min: 0,
      max: 500,
    };

    const component = shallowRender(props);

    expect(component).toMatchSnapshot();
  });

  it('render with min only', () => {
    const props = {
      min: 0,
    };

    const component = shallowRender(props);

    expect(component).toMatchSnapshot();
  });

  it('render with max only', () => {
    const props = {
      max: 500,
    };

    const component = shallowRender(props);

    expect(component).toMatchSnapshot();
  });

  it('render with precision of 1', () => {
    const props = {
      precision: 1,
    };

    const component = shallowRender(props);

    expect(component).toMatchSnapshot();
  });

  it('render with precision of 2', () => {
    const props = {
      precision: 2,
    };

    const component = shallowRender(props);

    expect(component).toMatchSnapshot();
  });

  it("render when can't refine", () => {
    const props = {
      canRefine: false,
      min: 0,
      max: 100,
      currentRefinement: {
        min: 10,
        max: 90,
      },
    };

    const component = shallowRender(props);

    expect(component).toMatchSnapshot();
  });

  describe('willMount', () => {
    it('expect to call canRefine from context when defined', () => {
      const props = {};
      const context = {
        canRefine: jest.fn(),
      };

      shallowRender(props, context);

      expect(context.canRefine).toHaveBeenCalledTimes(1);
    });

    it('expect to not throw when canRefine is not defined', () => {
      const props = {};
      const context = {};

      expect(() => shallowRender(props, context)).not.toThrow();
    });
  });

  describe('willReceiveProps', () => {
    it('expect to update state when props have changed ', () => {
      const props = {
        canRefine: false,
        currentRefinement: {
          min: 0,
          max: 100,
        },
      };

      const context = {
        canRefine: jest.fn(),
      };

      const wrapper = shallowRender(props, context);

      wrapper.setProps({
        canRefine: true,
        currentRefinement: {
          min: 10,
          max: 90,
        },
      });

      wrapper.update();

      expect(wrapper.state()).toEqual({
        from: 10,
        to: 90,
      });
    });

    it("expect to not update state when props don't have changed", () => {
      const props = {
        canRefine: true,
        currentRefinement: {
          min: 0,
          max: 100,
        },
      };

      const context = {
        canRefine: jest.fn(),
      };

      const wrapper = shallowRender(props, context);

      wrapper.setState({
        from: 10,
        to: 90,
      });

      wrapper.setProps({
        canRefine: true,
        currentRefinement: {
          min: 0,
          max: 100,
        },
      });

      wrapper.update();

      expect(wrapper.state()).toEqual({
        from: 10,
        to: 90,
      });
    });

    it('expect to call context canRefine when props changed', () => {
      const props = {
        canRefine: true,
        currentRefinement: {
          min: 0,
          max: 100,
        },
      };

      const context = {
        canRefine: jest.fn(),
      };

      const wrapper = shallowRender(props, context);

      wrapper.setProps({
        canRefine: false,
      });

      expect(context.canRefine).toHaveBeenCalledTimes(2);
    });

    it("expect to not call context canRefine when props don't have changed", () => {
      const props = {
        canRefine: true,
        currentRefinement: {
          min: 0,
          max: 100,
        },
      };

      const context = {
        canRefine: jest.fn(),
      };

      const wrapper = shallowRender(props, context);

      wrapper.setProps({
        canRefine: true,
      });

      expect(context.canRefine).toHaveBeenCalledTimes(1);
    });

    it('expect to not throw when canRefine is not defined', () => {
      const props = {};
      const context = {};
      const nextProps = {};

      const component = shallowRender(props, context);

      expect(() => component.setProps(nextProps)).not.toThrow();
    });
  });

  describe('onChange', () => {
    it('expect to update min onChange', () => {
      const props = {};
      const component = shallowRender(props);

      component.find('.ais-RangeInput__inputMin').simulate('change', {
        currentTarget: {
          value: 10,
        },
      });

      expect(component).toMatchSnapshot();
      expect(component.state()).toEqual({
        from: 10,
        to: '',
      });
    });

    it('expect to update max onChange', () => {
      const props = {};
      const component = shallowRender(props);

      component.find('.ais-RangeInput__inputMax').simulate('change', {
        currentTarget: {
          value: 490,
        },
      });

      expect(component).toMatchSnapshot();
      expect(component.state()).toEqual({
        from: '',
        to: 490,
      });
    });
  });

  describe('onSubmit', () => {
    it('expect to call refine onSubmit with values', () => {
      const props = {
        refine: jest.fn(),
      };

      const event = {
        preventDefault: jest.fn(),
      };

      const component = shallowRender(props);

      component.setState({
        from: 10,
        to: 490,
      });

      component.find('form').simulate('submit', event);

      expect(event.preventDefault).toHaveBeenCalled();
      expect(props.refine).toHaveBeenCalledWith({
        min: 10,
        max: 490,
      });
    });

    it('expect to not call refine with empty string', () => {
      const props = {
        refine: jest.fn(),
      };

      const event = {
        preventDefault: jest.fn(),
      };

      const component = shallowRender(props);

      component.find('form').simulate('submit', event);

      expect(event.preventDefault).toHaveBeenCalled();
      expect(props.refine).toHaveBeenCalledWith({
        min: '',
        max: '',
      });
    });
  });
});
