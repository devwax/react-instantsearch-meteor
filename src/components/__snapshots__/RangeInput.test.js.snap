// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeInput render with translations 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      SEPARATOR
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      SUBMIT
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput onChange expect to update max onChange 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value={490}
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput onChange expect to update min onChange 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value={10}
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput render when can't refine 1`] = `
<form
  className="ais-RangeInput__root ais-RangeInput__noRefinement"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={true}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput render with empty values 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput render with empty values when refinement is equal to min / max 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max={500}
        min={0}
        onChange={[Function]}
        placeholder={0}
        step={1}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max={500}
        min={0}
        onChange={[Function]}
        placeholder={500}
        step={1}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput render with max only 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput render with min / max 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max={500}
        min={0}
        onChange={[Function]}
        placeholder={0}
        step={1}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max={500}
        min={0}
        onChange={[Function]}
        placeholder={500}
        step={1}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput render with min only 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput render with precision of 1 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={0.1}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={0.1}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput render with precision of 2 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={0.01}
        type="number"
        value=""
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={0.01}
        type="number"
        value=""
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;

exports[`RawRangeInput render with refinement 1`] = `
<form
  className="ais-RangeInput__root"
  onSubmit={[Function]}
>
  <fieldset
    className="ais-RangeInput__fieldset"
    disabled={false}
  >
    <label
      className="ais-RangeInput__labelMin"
    >
      <input
        className="ais-RangeInput__inputMin"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value={10}
      />
    </label>
    <span
      className="ais-RangeInput__separator"
    >
      separator
    </span>
    <label
      className="ais-RangeInput__labelMax"
    >
      <input
        className="ais-RangeInput__inputMax"
        max=""
        min=""
        onChange={[Function]}
        placeholder=""
        step={1}
        type="number"
        value={490}
      />
    </label>
    <button
      className="ais-RangeInput__submit"
      type="submit"
    >
      submit
    </button>
  </fieldset>
</form>
`;
